In [2]:
import pandas as pd
import numpy as np

# DATASET 1

In [3]:
df1 = pd.read_csv('C://Users//pc//Downloads//adoptions//Dogs.csv', encoding = 'utf-8',sep=',')
df1.head()

id org_id                                                url type.x  \
0  46042150  NV163  https://www.petfinder.com/dog/harley-46042150/...    Dog   
1  46042002  NV163  https://www.petfinder.com/dog/biggie-46042002/...    Dog   
2  46040898   NV99  https://www.petfinder.com/dog/ziggy-46040898/n...    Dog   
3  46039877  NV202  https://www.petfinder.com/dog/gypsy-46039877/n...    Dog   
4  46039306  NV184  https://www.petfinder.com/dog/theo-46039306/nv...    Dog   

  species                   breed_primary breed_secondary  breed_mixed  \
0     Dog  American Staffordshire Terrier     Mixed Breed         True   
1     Dog                Pit Bull Terrier     Mixed Breed         True   
2     Dog                        Shepherd             NaN        False   
3     Dog             German Shepherd Dog             NaN        False   
4     Dog                       Dachshund             NaN        False   

   breed_unknown      color_primary  ... contact_city contact_state  \
0          False      White / Cream  ...    Las Vegas            NV   
1          False  Brown / Chocolate  ...    Las Vegas            NV   
2          False            Brindle  ...     Mesquite            NV   
3          False                NaN  ...      Pahrump            NV   
4          False                NaN  ...    Henderson            NV   

  contact_zip contact_country stateQ    accessed  type.y  \
0       89147              US  89009  2019-09-20     Dog   
1       89147              US  89009  2019-09-20     Dog   
2       89027              US  89009  2019-09-20     Dog   
3       89048              US  89009  2019-09-20     Dog   
4       89052              US  89009  2019-09-20     Dog   

                                         description  stay_duration  stay_cost  
0  Harley is not sure how he wound up at shelter ...             70     124.81  
1  6 year old Biggie has lost his home and really...             49     122.07  
2  Approx 2 years old.\n Did I catch your eye? I ...             87     281.51  
3                                                NaN             62     145.83  
4  Theo is a friendly dachshund mix who gets alon...             93     241.09  

[5 rows x 37 columns]

In [3]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58180 entries, 0 to 58179
Data columns (total 37 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               58180 non-null  int64  
 1   org_id           58180 non-null  object 
 2   url              58180 non-null  object 
 3   type.x           58180 non-null  object 
 4   species          58180 non-null  object 
 5   breed_primary    58180 non-null  object 
 6   breed_secondary  20821 non-null  object 
 7   breed_mixed      58180 non-null  bool   
 8   breed_unknown    58180 non-null  bool   
 9   color_primary    26134 non-null  object 
 10  color_secondary  12059 non-null  object 
 11  color_tertiary   1217 non-null   object 
 12  age              58180 non-null  object 
 13  sex              58180 non-null  object 
 14  size             58180 non-null  object 
 15  coat             27185 non-null  object 
 16  fixed            58180 non-null  bool   
 17  house_traine

In [4]:
df1.isnull().sum()

id                     0
org_id                 0
url                    0
type.x                 0
species                0
breed_primary          0
breed_secondary    37359
breed_mixed            0
breed_unknown          0
color_primary      32046
color_secondary    46121
color_tertiary     56963
age                    0
sex                    0
size                   0
coat               30995
fixed                  0
house_trained          0
declawed           58180
special_needs          0
shots_current          0
env_children       30153
env_dogs           23511
env_cats           38828
name                   0
status                 0
posted                 0
contact_city           0
contact_state          0
contact_zip           12
contact_country        0
stateQ                 0
accessed              33
type.y               640
description         8705
stay_duration          0
stay_cost              0
dtype: int64

In [5]:
df1.shape

(58180, 37)

# Piccola esplorazione del dataset
il primo dataset è formato da 58180 righe e 37 colonne, di queste 13 colonne contengono dei valori nulli: la domanda da porsi riguardo questi valori nulli è sono valori non esistenti o esistenti ma non inseriti nel DB? le colonne env_children, env_dogs, env_cats, contact_zip, color_primary, coat, ricadono nella seconda casistica, per cui si decide di eliminarne i valori nulli.

l = ['env_children', 'env_dogs', 'env_cats', 'contact_zip', 'color_primary', 'coat'] <br>
i = 0 <br>
for x in l:<br>
    df1 = df1[df1[x].notna()]<br>
    i=i+1<br>
    print('ho pulito il dataset per la volta numero:', i)<br>
print('ho pulito tutto il dataset di nostro interesse')<br>
<br>
for x in l: <br>
    conta = df1[x].isnull().sum()<br>
    print('il dataset per la colonna ', x,  ' possiede ', conta, ' valori nulli')<br>
    

In [8]:
df1.shape

(6697, 37)

il problema della rimozione di questi duplicati è che delle 58180 righe iniziali ne rimarrebbero solo 6697, per cui si da per buono a livello concettuale questa rimozione dei duplicati, ma non la si applica nella pipeline effettiva.

# DATASET 2

In [4]:
df2 = pd.read_csv('C://Users//pc//Downloads//adoptions//dogTravel.csv', encoding = 'utf-8',sep=',',index_col=0)
df2.head()

id contact_city contact_state  \
index                                        
0      44520267        Anoka            MN   
1      44698509    Groveland            FL   
2      45983838    Adamstown            MD   
3      44475904  Saint Cloud            MN   
4      43877389       Pueblo            CO   

                                             description        found  \
index                                                                   
0      Boris is a handsome mini schnauzer who made hi...     Arkansas   
1      Duke is an almost 2 year old Potcake from Abac...       Abacos   
2      Zac Woof-ron is a heartthrob movie star lookin...         Adam   
3      ~~Came in to the shelter as a transfer from an...      Adaptil   
4      Palang is such a sweetheart. She loves her peo...  Afghanistan   

         manual remove still_there  
index                               
0           NaN    NaN         NaN  
1       Bahamas    NaN         NaN  
2      Maryland    NaN         NaN  
3           NaN   True         NaN  
4           NaN    NaN         NaN

In [7]:
df2.isnull().sum()

id                  0
contact_city        0
contact_state       0
description         0
found               0
manual           4047
remove           4456
still_there      5875
dtype: int64

In [8]:
df2.shape

(6194, 8)

Il secondo dataset è composto da 8 colonne e 6194 righe: nonostante la folta presenza di valori null nel dataset ho deciso di non rimuovere le righe in quanto le informazioni nulle sono irrilevanti rispetto alle query da effettuare, inoltre una rimozione significherebbe la perdita di molte informazioni all'interno del dataset stesso

# DATASET 3

In [5]:
df3 = pd.read_csv('C://Users//pc//Downloads//adoptions//NST-EST2021-POP.csv', encoding = 'utf-8',sep=',')
df3.head()

Alabama   5.024.279
0      Alaska     733.391
1     Arizona   7.151.502
2    Arkansas   3.011.524
3  California  39.538.223
4    Colorado   5.773.714

In [6]:
df3.columns = ['state', 'population']

In [7]:
type(df3['population'][0])

str

In [8]:
new_row = pd.Series(["Alabama", "5.024.279"], index=['state', 'population'])

In [9]:
df3 = df3.append(new_row, ignore_index=True)

In [10]:
df3 = df3.sort_values('state')

In [11]:
df3.head()

state  population
50     Alabama   5.024.279
0       Alaska     733.391
1      Arizona   7.151.502
2     Arkansas   3.011.524
3   California  39.538.223

In [12]:
df3.isnull().sum()

state         0
population    0
dtype: int64

In [13]:
df3.shape

(51, 2)

Il terzo dataset è composto da 51 righe e 2 colonne: questo non presenta valori nulli

# QUERY 1
## Extract all dogs with status that is not adoptable
chiedi a fabri se serve qualche prepro

Per effettuare la prima query è stato effettuato un loc su status, cercando selezionando tutte le righe nelle quali questo non è uguale ad "adoptabile"

In [47]:
query1 = df1.loc[df1['status'] != 'adoptable']
query1.head()

id org_id                                                url  \
644    41330726  NV173  https://www.petfinder.com/dog/gunther-gunny-41...   
5549   38169117  AZ414  https://www.petfinder.com/dog/annabelle-annie-...   
10888  45833989   NY98  https://www.petfinder.com/dog/pepper-courtesy-...   
11983  45515547   NY98  https://www.petfinder.com/dog/cooper-courtesy-...   
12495  45294115   NY98  https://www.petfinder.com/dog/daisy-courtesy-l...   

      type.x species        breed_primary   breed_secondary  breed_mixed  \
644      Dog     Dog  German Shepherd Dog               NaN        False   
5549     Dog     Dog                Boxer  Pit Bull Terrier         True   
10888    Dog     Dog               Beagle               NaN        False   
11983    Dog     Dog          Mixed Breed               NaN        False   
12495    Dog     Dog         Basset Hound               NaN        False   

       breed_unknown      color_primary  ... contact_city contact_state  \
644            False                NaN  ...           NV         89146   
5549           False              Black  ...           AZ         85249   
10888          False                NaN  ...           NY         12220   
11983          False                NaN  ...           NY         12220   
12495          False  Brown / Chocolate  ...           NY         12220   

      contact_zip contact_country      stateQ accessed  type.y  \
644            US           89009  2019-09-20      NaN     Dog   
5549           US              AZ  2019-09-20      NaN     Dog   
10888          US              CT  2019-09-20      NaN     Dog   
11983          US              CT  2019-09-20      NaN     Dog   
12495          US              CT  2019-09-20      NaN     Dog   

                                             description  stay_duration  \
644    Meet handsome 3 year old Gunther. Gunther came...            108   
5549   You can fill out an adoption application onlin...             80   
10888  This is Pepper. He is a 15 year old tri-color ...             86   
11983  Cooper is 13 years old, but according to a ver...            105   
12495  •Basset Hound, female, •10 years \n\nDelig...             57   

       stay_cost  
644       256.88  
5549      130.77  
10888     180.70  
11983     400.82  
12495      82.61  

[5 rows x 37 columns]

# QUERY 2
## For each (primary) breed, determine the number of dogs

Per effettuare la seconda query è stato effettuato un groupby sulla razza primaria e poi un count per ogni razza raggruppata

In [48]:
query2 = df1.groupby(['breed_primary'], sort=True)['breed_primary'].count()

In [49]:
type(query2)

pandas.core.series.Series

In [51]:
df_query2 = query2.to_frame()

In [52]:
df_query2.columns = ['count']

In [53]:
df_query2

count
breed_primary                           
Affenpinscher                         17
Afghan Hound                           4
Airedale Terrier                      19
Akbash                                 3
Akita                                181
...                                  ...
Wirehaired Pointing Griffon            1
Wirehaired Terrier                    60
Xoloitzcuintli / Mexican Hairless     11
Yellow Labrador Retriever            158
Yorkshire Terrier                    360

[216 rows x 1 columns]

# QUERY 3
## For each (primary) breed, determine the ratio between the number of dogs of Mixed Breed and those not of Mixed Breed. Hint: look at the secondary_breed

Vado a creare un dataframe per poter costruire la colonna "ratio" che per ogni cane conterrà il rapporto tra cani con razza mista e cani senza razza mista.

In [3]:
query3 = df1[['breed_primary', 'breed_secondary','breed_mixed']]
query3.head()

breed_primary breed_secondary  breed_mixed
0  American Staffordshire Terrier     Mixed Breed         True
1                Pit Bull Terrier     Mixed Breed         True
2                        Shepherd             NaN        False
3             German Shepherd Dog             NaN        False
4                       Dachshund             NaN        False

Mi occupo di creare tre dataframe: uno che tenga conto del numero di cani per ogni razza che possiedono una razza secondaria, uno che tenga conto del numero di cani per ogni razza ed uno per ogni razza che non possiede una razza secondaria (ottenuto per differenza tra il secondo ed il primo dataframe).

In [4]:
query = query3.groupby(by='breed_primary')
not_null = query['breed_secondary'].count()
not_null

breed_primary
Affenpinscher                          2
Afghan Hound                           1
Airedale Terrier                       9
Akbash                                 0
Akita                                 52
                                    ... 
Wirehaired Pointing Griffon            0
Wirehaired Terrier                    18
Xoloitzcuintli / Mexican Hairless      4
Yellow Labrador Retriever             62
Yorkshire Terrier                    129
Name: breed_secondary, Length: 216, dtype: int64

In [7]:
not_null_df = not_null.to_frame()

In [16]:
not_null_df.columns = ['mixed_breed']
not_null_df.head()

mixed_breed
breed_primary                
Affenpinscher               2
Afghan Hound                1
Airedale Terrier            9
Akbash                      0
Akita                      52

In [5]:
count = query3.groupby(['breed_primary'], sort=True)['breed_primary'].count()
count

breed_primary
Affenpinscher                         17
Afghan Hound                           4
Airedale Terrier                      19
Akbash                                 3
Akita                                181
                                    ... 
Wirehaired Pointing Griffon            1
Wirehaired Terrier                    60
Xoloitzcuintli / Mexican Hairless     11
Yellow Labrador Retriever            158
Yorkshire Terrier                    360
Name: breed_primary, Length: 216, dtype: int64

In [14]:
count_df = count.to_frame()
count_df.columns = ['count']
count_df.head()

count
breed_primary          
Affenpinscher        17
Afghan Hound          4
Airedale Terrier     19
Akbash                3
Akita               181

In [6]:
null = count - not_null
null

breed_primary
Affenpinscher                         15
Afghan Hound                           3
Airedale Terrier                      10
Akbash                                 3
Akita                                129
                                    ... 
Wirehaired Pointing Griffon            1
Wirehaired Terrier                    42
Xoloitzcuintli / Mexican Hairless      7
Yellow Labrador Retriever             96
Yorkshire Terrier                    231
Length: 216, dtype: int64

In [17]:
null_df = null.to_frame()
null_df.columns = ['not_mixed_breed']
null_df.head()

not_mixed_breed
breed_primary                    
Affenpinscher                  15
Afghan Hound                    3
Airedale Terrier               10
Akbash                          3
Akita                         129

In [18]:
df = count_df.merge(not_null_df, left_index=True, right_index=True).merge(null_df, left_index=True, right_index=True)
df.head()

count  mixed_breed  not_mixed_breed
breed_primary                                        
Affenpinscher        17            2               15
Afghan Hound          4            1                3
Airedale Terrier     19            9               10
Akbash                3            0                3
Akita               181           52              129

In [19]:
df['ratio'] = df['mixed_breed']/df['not_mixed_breed']

In [20]:
df.head()

count  mixed_breed  not_mixed_breed     ratio
breed_primary                                                  
Affenpinscher        17            2               15  0.133333
Afghan Hound          4            1                3  0.333333
Airedale Terrier     19            9               10  0.900000
Akbash                3            0                3  0.000000
Akita               181           52              129  0.403101

# QUERY 4
## For each (primary) breed, determine the earliest and the latest posted timestamp.

per realizzare questa query ho innanzitutto estrapolato le due colonne di interesse (breed primary e posted), per poi chiedre, per ogni razza, la più recente e la meno recente data in memoria, creando così un fataframe che per ogni razza possiede la la più recente e la meno decente data in posted 

In [6]:
df4 = df1[['breed_primary','posted']]
df4.head()

breed_primary                    posted
0  American Staffordshire Terrier  2019-09-20T16:37:59+0000
1                Pit Bull Terrier  2019-09-20T16:24:57+0000
2                        Shepherd  2019-09-20T14:10:11+0000
3             German Shepherd Dog  2019-09-20T10:08:22+0000
4                       Dachshund  2019-09-20T06:48:30+0000

In [7]:
df4 = df4.drop(df4[df4['posted'] == "Nashville"].index)
df4 = df4.drop(df4[df4['posted'] == "Seattle"].index)

In [8]:
last = df4.sort_values(['breed_primary','posted']).groupby('breed_primary').last()
last.head()

posted
breed_primary                             
Affenpinscher     2019-09-14T10:10:51+0000
Afghan Hound      2019-07-27T00:38:48+0000
Airedale Terrier  2019-09-19T18:40:39+0000
Akbash            2019-08-23T17:11:04+0000
Akita             2019-09-20T15:19:57+0000

In [9]:
first= df4.sort_values(['breed_primary','posted']).groupby('breed_primary').first()
first.head()

posted
breed_primary                             
Affenpinscher     2012-03-08T10:27:33+0000
Afghan Hound      2017-06-29T23:28:51+0000
Airedale Terrier  2014-06-13T12:59:36+0000
Akbash            2019-07-21T00:35:59+0000
Akita             2012-03-03T09:31:08+0000

In [10]:
final = first.merge(last, on='breed_primary')
final.columns = ['earliest', 'latest']
final.head()

earliest                    latest
breed_primary                                                       
Affenpinscher     2012-03-08T10:27:33+0000  2019-09-14T10:10:51+0000
Afghan Hound      2017-06-29T23:28:51+0000  2019-07-27T00:38:48+0000
Airedale Terrier  2014-06-13T12:59:36+0000  2019-09-19T18:40:39+0000
Akbash            2019-07-21T00:35:59+0000  2019-08-23T17:11:04+0000
Akita             2012-03-03T09:31:08+0000  2019-09-20T15:19:57+0000

# QUERY 5
## For each state, compute the sex imbalance, that is the difference between male and female dogs. In which state this imbalance is largest?

per effettuare questa query ho estrapolato per ogni contact state il numero di maschi e femmine, per poi calcolarne la differenza.

In [12]:
sex = df1[['contact_state','sex']]
male = sex.groupby('contact_state')['sex'].apply(lambda x: (x=='Male').sum()).reset_index(name = 'Male count')
male

contact_state  Male count
0          12220           2
1          12477           1
2          17325           1
3          19053           0
4          19063           0
..           ...         ...
70            VT         276
71            WA         682
72            WI         277
73            WV         333
74            WY          31

[75 rows x 2 columns]

In [13]:
female = sex.groupby('contact_state')['sex'].apply(lambda x: (x=='Female').sum()).reset_index(name = 'female count')
female

contact_state  female count
0          12220             1
1          12477             1
2          17325             1
3          19053             1
4          19063             1
..           ...           ...
70            VT           234
71            WA           595
72            WI           265
73            WV           232
74            WY            21

[75 rows x 2 columns]

In [15]:
C = male.merge(female, on='contact_state')
C.columns = ['contact state', 'Male count', 'Female count']
C['difference'] = C['Male count'] - C['Female count']
C

contact state  Male count  Female count  difference
0          12220           2             1           1
1          12477           1             1           0
2          17325           1             1           0
3          19053           0             1          -1
4          19063           0             1          -1
..           ...         ...           ...         ...
70            VT         276           234          42
71            WA         682           595          87
72            WI         277           265          12
73            WV         333           232         101
74            WY          31            21          10

[75 rows x 4 columns]

In [23]:
query5 = C.sort_values('difference',  ascending=False)

In [33]:
query5.head()

contact state  Male count  Female count  difference
58            OH        1439          1231         208
37            IN        1027           850         177
69            VA        1608          1450         158
42            MD         814           679         135
66            TN         944           825         119

# QUERY 6
## For each pair (age, size), determine the average duration of the stay and the average cost of stay.

per effettuare questa query ho innanzitutto estratto le colonne di interesse, andando poi per ogni coppia età e taglia a calcolare la media della durata e del costo.

In [14]:
query6 = df1[['age', 'size', 'stay_duration', 'stay_cost']]
query6.head()

age    size  stay_duration  stay_cost
0  Senior  Medium             70     124.81
1   Adult   Large             49     122.07
2   Adult   Large             87     281.51
3    Baby   Large             62     145.83
4   Young   Small             93     241.09

In [15]:
mean = query6.groupby(['age', 'size'])['stay_duration', 'stay_cost'].mean().reset_index()
mean.columns= ['age', 'size', 'stay duration average', 'stay cost average']
mean.head()

age         size  stay duration average  stay cost average
0  Adult  Extra Large              89.015414         232.591561
1  Adult        Large              89.531943         238.661141
2  Adult       Medium              89.421036         238.258977
3  Adult        Small              89.407479         238.974838
4   Baby  Extra Large              87.032967         237.180879

# QUERY 7
##  Find the dogs involved in at least 3 travels. Also list the breed of those dogs.

per questa query ho recuperato dal secondo dataset a disposizione, il numero di viaggi effettuati da ogni cane (raggruppando per id), successivamente ho recuperato dal primo dataset id e razza primaria di ogni cane: infine per creare il dataframe finale ho effettuato una left join tra le due tabelle.

In [36]:
t = df2[df2['id'].map(df2['id'].value_counts()) > 2]
travels = t.groupby(['id'])['id'].count().reset_index(name = 'numero_viaggi')
travels.head()

id  numero_viaggi
0  16657005              4
1  20905974              5
2  24894870              4
3  24894894              4
4  33218331              7

In [38]:
r = df1[['id', 'breed_primary']]
breed = travels.merge(r, on='id', how='left')
breed.head()

id  numero_viaggi     breed_primary
0  16657005              4  Pit Bull Terrier
1  20905974              5         Chow Chow
2  24894870              4             Hound
3  24894894              4             Hound
4  33218331              7  Alaskan Malamute

# QUERY8
## Fix the travels table so that the correct state is computed from the manual and the found fields. If manual is not missing, then it overrides what is stored in found.

Per realizzare l'ottava query ho estrapolato le tre colonne di nostro interesse: dopo di che ho sovrascritto in found il valore trovato in manual per tutte le righe non nulle.

In [16]:
df8 = df2[['id', 'found', 'manual']]
df8.head()

id        found    manual
index                                 
0      44520267     Arkansas       NaN
1      44698509       Abacos   Bahamas
2      45983838         Adam  Maryland
3      44475904      Adaptil       NaN
4      43877389  Afghanistan       NaN

In [17]:
df8['found'].isnull().sum()

0

In [18]:
df8['found'].mask(df8['manual'].notnull() == True, df8['manual'], inplace=True)

In [19]:
df8.head()

id        found    manual
index                                 
0      44520267     Arkansas       NaN
1      44698509      Bahamas   Bahamas
2      45983838     Maryland  Maryland
3      44475904      Adaptil       NaN
4      43877389  Afghanistan       NaN

# QUERY 9
## For each state, compute the ratio between the number of travels and the population.

Per realizzare la nona query effettuo un merge tra il secondo e il terzo dataset a disposzione: da questo costruisco un dataframe che raggruppando per stato tiene memoria di quanti viaggi sono stati effettuati, andandolo poi a mergiare con il terzo dataframe. 
una volta fatto ciò, realizzo il calcolo del rapporto.

In [20]:
df = df2[['id', 'found']]
df.columns = ['id', 'state']
df9 = df3.merge(df, on='state', how='left') 

In [21]:
df9.head()

state population          id
0  Alabama  5.024.279  45987719.0
1  Alabama  5.024.279  45943086.0
2  Alabama  5.024.279  45942459.0
3  Alabama  5.024.279  45942229.0
4  Alabama  5.024.279  45941983.0

In [22]:
a = df9.groupby(['state'], sort=True)['state'].count() #ottengo il numero di viaggi per ogni stato

In [23]:
prova = a.to_frame()

In [24]:
prova.columns = ['travel_count']

In [25]:
prova.head()

travel_count
state                   
Alabama              322
Alaska                 1
Arizona               18
Arkansas              35
California           130

In [26]:
df_appoggio = df9[['state', 'population']]
df_appoggio.head()

state population
0  Alabama  5.024.279
1  Alabama  5.024.279
2  Alabama  5.024.279
3  Alabama  5.024.279
4  Alabama  5.024.279

In [27]:
final = prova.merge(df3, on='state')  
final.head() #creato così il dataframe che possiede la popolazione e il numero di viaggi per ogni stato

state  travel_count  population
0     Alabama           322   5.024.279
1      Alaska             1     733.391
2     Arizona            18   7.151.502
3    Arkansas            35   3.011.524
4  California           130  39.538.223

In [28]:
final['population'] = final['population'].str.replace('.', '') #necessario rimuovere i punti per la trasformazione da stringa a numero

In [29]:
final['population'] = pd.to_numeric(final['population'])
final.head()

state  travel_count  population
0     Alabama           322     5024279
1      Alaska             1      733391
2     Arizona            18     7151502
3    Arkansas            35     3011524
4  California           130    39538223

In [30]:
final['ratio'] = final['travel_count']/final['population']
final.head()

state  travel_count  population     ratio
0     Alabama           322     5024279  0.000064
1      Alaska             1      733391  0.000001
2     Arizona            18     7151502  0.000003
3    Arkansas            35     3011524  0.000012
4  California           130    39538223  0.000003

# QUERY 10
## For each dog, compute the number of days from the posted day to the day of last access.

Per effettuare la decima query ho estratto le colonne rilevanti dal primo dataset per poi andare a pulire le colonne "posted" e "accessed": per la seconda è stat necessaria una semplice conversione, mentre per la prima per poter effettuarla ho dovuto eseguire degli step di pulizia (rimozione caratteri indesiderati, rimozione righe non contenenti date) andando successivamente a realizzare la differenza richiesta sulle due date ottenute

In [31]:
df10 = df1[['id', 'posted', 'accessed']]
df10.head()

id                    posted    accessed
0  46042150  2019-09-20T16:37:59+0000  2019-09-20
1  46042002  2019-09-20T16:24:57+0000  2019-09-20
2  46040898  2019-09-20T14:10:11+0000  2019-09-20
3  46039877  2019-09-20T10:08:22+0000  2019-09-20
4  46039306  2019-09-20T06:48:30+0000  2019-09-20

In [32]:
df10['accessed'] = pd.to_datetime(df10['accessed'], format='%Y-%m-%d')

### pulizia posted

#### 1. rimuovo caratteri indesiderati dalla data

In [33]:
pat = f'(?i)^(.*)({"|".join("T")}).*$'
df10['posted'] = df10['posted'].str.replace(pat, r'\1\2', regex=True)
df10['posted'] = df10['posted'].replace({'T':''}, regex=True)

In [34]:
df10['posted']

0        2019-09-20
1        2019-09-20
2        2019-09-20
3        2019-09-20
4        2019-09-20
            ...    
58175    2019-05-03
58176    2019-04-13
58177    2018-09-27
58178    2018-09-12
58179    2018-09-03
Name: posted, Length: 58180, dtype: object

#### 2. trasformo la data da stringa a date time tenendo memoria di quelle righe che non contengono date

In [35]:
pulire = []

for x in range(0,len(df10['posted'])):
    try:
        df10['posted'][x] = pd.to_datetime(df10['posted'][x], format='%Y-%m-%d')
    except ValueError:
        pulire.append(df10['posted'][x])
    print('effettuata la procedura per il valore', x)

effettuata la procedura per il valore 0
effettuata la procedura per il valore 1
effettuata la procedura per il valore 2
effettuata la procedura per il valore 3
effettuata la procedura per il valore 4
effettuata la procedura per il valore 5
effettuata la procedura per il valore 6
effettuata la procedura per il valore 7
effettuata la procedura per il valore 8
effettuata la procedura per il valore 9
effettuata la procedura per il valore 10
effettuata la procedura per il valore 11
effettuata la procedura per il valore 12
effettuata la procedura per il valore 13
effettuata la procedura per il valore 14
effettuata la procedura per il valore 15
effettuata la procedura per il valore 16
effettuata la procedura per il valore 17
effettuata la procedura per il valore 18
effettuata la procedura per il valore 19
effettuata la procedura per il valore 20
effettuata la procedura per il valore 21
effettuata la procedura per il valore 22
effettuata la procedura per il valore 23
effettuata la procedura pe

In [36]:
pulire

['Las Vegas',
 'Chandler',
 'Albany',
 'Albany',
 'Albany',
 'Saugert',
 'Saugert',
 'Brist',
 'Silver Spring',
 'Gett',
 'Gett',
 'Media',
 'Feast',
 'Paris',
 'Mooresville',
 'Ross',
 'Ross',
 'Jeffersonville',
 'Cincinnat',
 'Paoli',
 'lake charles',
 'Albuquerque',
 'Cookeville',
 'Nashville',
 'Rust',
 'Midlot',
 'Seatt',
 'Seatt',
 'Seatt',
 'Seatt',
 'Seatt',
 'Seatt',
 'Seatt']

In [37]:
df10 = df10[~df10['posted'].isin(pulire)] #rimuovo gli elementi indesiderati

In [38]:
df10['diff_days'] = (df10['accessed'] - df10['posted']) / np.timedelta64(1, 'D')

In [39]:
df10

id               posted   accessed  diff_days
0      46042150  2019-09-20 00:00:00 2019-09-20        0.0
1      46042002  2019-09-20 00:00:00 2019-09-20        0.0
2      46040898  2019-09-20 00:00:00 2019-09-20        0.0
3      46039877  2019-09-20 00:00:00 2019-09-20        0.0
4      46039306  2019-09-20 00:00:00 2019-09-20        0.0
...         ...                  ...        ...        ...
58175  44605893  2019-05-03 00:00:00 2019-09-20      140.0
58176  44457061  2019-04-13 00:00:00 2019-09-20      160.0
58177  42865848  2018-09-27 00:00:00 2019-09-20      358.0
58178  42734734  2018-09-12 00:00:00 2019-09-20      373.0
58179  42663515  2018-09-03 00:00:00 2019-09-20      382.0

[58147 rows x 4 columns]

In [40]:
df10.duplicated(subset=['id']).sum()

0

# QUERY 11
##  Partition the dogs according to the number of weeks from the posted day to the day of last access.

### continuo a lavorare su df10

lavorando sullo stesso dataframe utilizzato per la query 10, si effettua lo stesso calcolo badando però al numero di settimane

In [41]:
df11 = df10.drop(['diff_days'], axis=1)
df11.head()

id               posted   accessed
0  46042150  2019-09-20 00:00:00 2019-09-20
1  46042002  2019-09-20 00:00:00 2019-09-20
2  46040898  2019-09-20 00:00:00 2019-09-20
3  46039877  2019-09-20 00:00:00 2019-09-20
4  46039306  2019-09-20 00:00:00 2019-09-20

In [42]:
df11['diff_weeks'] = round((df11['accessed'] - df11['posted']) / np.timedelta64(1, 'W'))

In [43]:
df11

id               posted   accessed  diff_weeks
0      46042150  2019-09-20 00:00:00 2019-09-20         0.0
1      46042002  2019-09-20 00:00:00 2019-09-20         0.0
2      46040898  2019-09-20 00:00:00 2019-09-20         0.0
3      46039877  2019-09-20 00:00:00 2019-09-20         0.0
4      46039306  2019-09-20 00:00:00 2019-09-20         0.0
...         ...                  ...        ...         ...
58175  44605893  2019-05-03 00:00:00 2019-09-20        20.0
58176  44457061  2019-04-13 00:00:00 2019-09-20        23.0
58177  42865848  2018-09-27 00:00:00 2019-09-20        51.0
58178  42734734  2018-09-12 00:00:00 2019-09-20        53.0
58179  42663515  2018-09-03 00:00:00 2019-09-20        55.0

[58147 rows x 4 columns]

# QUERY 12
## Find for duplicates in the dogs dataset. Two records are duplicates if they have (1) same breeds and sex, and (2) they share at least 90% of the words in the description field. Extra points if you find and implement a more refined for determining if two rows are duplicates.

Per poter effettuare la query numero 12 è stato necessario innanzitutto pulire il dataset, rimuovendo tutte quelle righe che contenevano un valore nullo per la descrizione.
Successivamente vi è un preprocessing per uniformare i testi e pulili da caratteri inutili ai fini del calcolo (link, punteggiatura, singoli caratteri particolari e minuscolo).

Il dataset viene inoltre ordinato in ordine crescente sulla base del proprio id.

Per poter ritrovare tutte le coppie di cani che condividono lo stesso sesso e la stessa razza viene effettuata una self join sul dataframe stesso sulle colonne "breed_primary", "breed_secondary" e "sex": il risultato sarà un dataframe con 7 colonne che descriveranno id del primo cane, descrizione del primo cane, id secondo cane, id secondo cane, sesso e razza primaria e secondaria. Su questo dataframe si specifica la condizione per cui il primo id deve sempre essere diverso dal secondo id ed inoltre il primo id deve essere minore del secondo, così da non avere righe duplicate con gli id invertiti.

Dunque si effettua il calcolo della similarità che per ogni riga calcola la percentuale di parole in comune come l'intersezione delle due parole fratto l'unione. Viene successivamente filtrato il dataframe andando ad ottenere quello definitivo con la percentuale superiore o pari a 0.90.

In [95]:
df1['description'].isnull().sum()

8705

In [96]:
df1 = df1.dropna(subset=['description'])#rimuovo le descrizioni nulle

In [97]:
df1['description'].isnull().sum() 

0

In [98]:
import nltk
from nltk.tokenize import word_tokenize
import string
import re

# Scarichiamo i dati necessari per la pulizia
nltk.download('punkt')
def morandi(text):
    """
    Pulisce il testo delle descrizioni.
    """
    # Rimuoviamo i link
    text = re.sub(r'http\S+', '', text)
    
    # Rimuove la punteggiatura
    text = text.translate(str.maketrans('', '', string.punctuation))
    
    # Rimuoviamo le singole lettere speciali
    text = re.sub(r'\b\w\b', '', text)
    
    # Rimuoviamo i caratteri indesiderati
    text = re.sub(r'[^\w\s]', '', text)
    
    # Rimuoviamo i caratteri di tabulazione e a capo
    text = re.sub(r'[\n\t]', '', text)
    
    # testo in minuscolo
    text = text.lower()

    return text

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\pc\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [99]:
df1['description'] = df1['description'].apply(morandi) #pulisco il testo da caratteri che lo appesantiscono (no rimozione stopword)

##### Ordinamento dei valori sulla base dell'id , in ordine crescente 

In [100]:
df1 = df1.sort_values('id')

### 1. SELF JOIN SU BREEDS E SEX 

In [101]:
appoggio = df1[['id', 'breed_primary', 'breed_secondary','sex', 'description']] 

In [102]:
joined_df = pd.merge(appoggio, appoggio, on=['breed_primary', 'breed_secondary','sex']) #self-join su breed primary e sesso

In [103]:
joined_df

id_x             breed_primary  \
0           604115                    Beagle   
1           604115                    Beagle   
2           604115                    Beagle   
3           604115                    Beagle   
4           604115                    Beagle   
...            ...                       ...   
29152404  46041795                  Shepherd   
29152405  46041816                  Shepherd   
29152406  46041816                  Shepherd   
29152407  46042053  Treeing Walker Coonhound   
29152408  46042101        Bluetick Coonhound   

                              breed_secondary     sex  \
0                                       Hound  Female   
1                                       Hound  Female   
2                                       Hound  Female   
3                                       Hound  Female   
4                                       Hound  Female   
...                                       ...     ...   
29152404                Catahoula Leopard Dog  Female   
29152405                Catahoula Leopard Dog  Female   
29152406                Catahoula Leopard Dog  Female   
29152407                          Plott Hound    Male   
29152408  Australian Cattle Dog / Blue Heeler  Female   

                                              description_x      id_y  \
0          save  life adopt dont buy     maricopa county...    604115   
1          save  life adopt dont buy     maricopa county...  35840400   
2          save  life adopt dont buy     maricopa county...  40869429   
3          save  life adopt dont buy     maricopa county...  42681874   
4          save  life adopt dont buy     maricopa county...  44451363   
...                                                     ...       ...   
29152404  trix  is   week old beautiful female shepherdc...  46041816   
29152405  lux  is   week old beautiful female shepherdca...  46041795   
29152406  lux  is   week old beautiful female shepherdca...  46041816   
29152407  primary color black secondary color brown brin...  46042053   
29152408  ophelia  is  blue tick houndblue heeler mix bo...  46042101   

                                              description_y  
0          save  life adopt dont buy     maricopa county...  
1         beautiful daisy her  siblings and mama mia wer...  
2         if you are interested in adopting this dog ple...  
3         im kasey   year old beagle mix  was rescued fr...  
4         please read descriptions in fullmeet jessica s...  
...                                                     ...  
29152404  lux  is   week old beautiful female shepherdca...  
29152405  trix  is   week old beautiful female shepherdc...  
29152406  lux  is   week old beautiful female shepherdca...  
29152407  primary color black secondary color brown brin...  
29152408  ophelia  is  blue tick houndblue heeler mix bo...  

[29152409 rows x 7 columns]

In [104]:
result =joined_df[(joined_df['id_x'] != joined_df['id_y']) & (joined_df['id_x'] < joined_df['id_y'])].copy()
#seleziona solo le righe in cui gli ID non sono uguali e l'ID nella prima posizione è minore dell'ID nella seconda posizione. Questo esclude le coppie di righe con gli stessi ID ma in posizioni inverse.

In [105]:
result

id_x                        breed_primary  \
1           604115                               Beagle   
2           604115                               Beagle   
3           604115                               Beagle   
4           604115                               Beagle   
5           604115                               Beagle   
...            ...                                  ...   
29152364  46036308  Australian Cattle Dog / Blue Heeler   
29152365  46036308  Australian Cattle Dog / Blue Heeler   
29152368  46036880  Australian Cattle Dog / Blue Heeler   
29152394  46039978                                Husky   
29152404  46041795                             Shepherd   

                breed_secondary     sex  \
1                         Hound  Female   
2                         Hound  Female   
3                         Hound  Female   
4                         Hound  Female   
5                         Hound  Female   
...                         ...     ...   
29152364       American Bulldog    Male   
29152365       American Bulldog    Male   
29152368       American Bulldog    Male   
29152394                Terrier  Female   
29152404  Catahoula Leopard Dog  Female   

                                              description_x      id_y  \
1          save  life adopt dont buy     maricopa county...  35840400   
2          save  life adopt dont buy     maricopa county...  40869429   
3          save  life adopt dont buy     maricopa county...  42681874   
4          save  life adopt dont buy     maricopa county...  44451363   
5          save  life adopt dont buy     maricopa county...  44659739   
...                                                     ...       ...   
29152364  hi iâ riley iâ     year old neutered male aust...  46036880   
29152365  hi iâ riley iâ     year old neutered male aust...  46041704   
29152368  hi iâ riley iâ     year old neutered male aust...  46041704   
29152394   pre adoption application must be completed fo...  46039983   
29152404  trix  is   week old beautiful female shepherdc...  46041816   

                                              description_y  
1         beautiful daisy her  siblings and mama mia wer...  
2         if you are interested in adopting this dog ple...  
3         im kasey   year old beagle mix  was rescued fr...  
4         please read descriptions in fullmeet jessica s...  
5         please contact pet informationpethelpersincorg...  
...                                                     ...  
29152364  hi iâ riley iâ     year old neutered male aust...  
29152365  hi iâ riley iâ     year old neutered male aust...  
29152368  hi iâ riley iâ     year old neutered male aust...  
29152394   pre adoption application must be completed fo...  
29152404  lux  is   week old beautiful female shepherdca...  

[14551467 rows x 7 columns]

#### similarità: intersezione delle parole tra le due descrizioni / unione delle parole delle due descrizioni (tempo: 17 minuti)

In [106]:
import time

In [107]:
def percentuale_parole_comuni(str_1, str_2):
    # dividi le stringhe in parole
    parole_1 = set(str_1.split())
    parole_2 = set(str_2.split())
    # calcola la percentuale di parole comuni
    percentuale = len(parole_1.intersection(parole_2)) / len(parole_1.union(parole_2))
    return percentuale
#semplice calcolo che si basa sul rapporto tra l'intersezione e l'unione di due parole (più o meno un'ora per il completamento)
#ricorda che si è spento il pc

In [108]:
start_time = time.time()

result['percentuale_parole_comuni'] = result.apply(lambda row: percentuale_parole_comuni(str(row['description_x']), str(row['description_y'])), axis=1)

end_time = time.time()

print("Tempo di esecuzione:", end_time - start_time, "secondi")

Tempo di esecuzione: 1029.344788312912 secondi


In [109]:
result 

id_x                        breed_primary  \
1           604115                               Beagle   
2           604115                               Beagle   
3           604115                               Beagle   
4           604115                               Beagle   
5           604115                               Beagle   
...            ...                                  ...   
29152364  46036308  Australian Cattle Dog / Blue Heeler   
29152365  46036308  Australian Cattle Dog / Blue Heeler   
29152368  46036880  Australian Cattle Dog / Blue Heeler   
29152394  46039978                                Husky   
29152404  46041795                             Shepherd   

                breed_secondary     sex  \
1                         Hound  Female   
2                         Hound  Female   
3                         Hound  Female   
4                         Hound  Female   
5                         Hound  Female   
...                         ...     ...   
29152364       American Bulldog    Male   
29152365       American Bulldog    Male   
29152368       American Bulldog    Male   
29152394                Terrier  Female   
29152404  Catahoula Leopard Dog  Female   

                                              description_x      id_y  \
1          save  life adopt dont buy     maricopa county...  35840400   
2          save  life adopt dont buy     maricopa county...  40869429   
3          save  life adopt dont buy     maricopa county...  42681874   
4          save  life adopt dont buy     maricopa county...  44451363   
5          save  life adopt dont buy     maricopa county...  44659739   
...                                                     ...       ...   
29152364  hi iâ riley iâ     year old neutered male aust...  46036880   
29152365  hi iâ riley iâ     year old neutered male aust...  46041704   
29152368  hi iâ riley iâ     year old neutered male aust...  46041704   
29152394   pre adoption application must be completed fo...  46039983   
29152404  trix  is   week old beautiful female shepherdc...  46041816   

                                              description_y  \
1         beautiful daisy her  siblings and mama mia wer...   
2         if you are interested in adopting this dog ple...   
3         im kasey   year old beagle mix  was rescued fr...   
4         please read descriptions in fullmeet jessica s...   
5         please contact pet informationpethelpersincorg...   
...                                                     ...   
29152364  hi iâ riley iâ     year old neutered male aust...   
29152365  hi iâ riley iâ     year old neutered male aust...   
29152368  hi iâ riley iâ     year old neutered male aust...   
29152394   pre adoption application must be completed fo...   
29152404  lux  is   week old beautiful female shepherdca...   

          percentuale_parole_comuni  
1                          0.089855  
2                          0.081023  
3                          0.089286  
4                          0.108374  
5                          0.097130  
...                             ...  
29152364                   0.991342  
29152365                   1.000000  
29152368                   0.991342  
29152394                   1.000000  
29152404                   0.978495  

[14551467 rows x 8 columns]

In [110]:
result_filtrato = result[result['percentuale_parole_comuni'] >= 0.9] #filtro lasciando solo le descrizioni con il 90% delle parole in comune

In [111]:
result_filtrato

id_x                        breed_primary  \
20295     34418553                          Rat Terrier   
23108     44990257                          Rat Terrier   
25630     45864886                          Rat Terrier   
27764     46015088                          Rat Terrier   
28249     46033636                          Rat Terrier   
...            ...                                  ...   
29152364  46036308  Australian Cattle Dog / Blue Heeler   
29152365  46036308  Australian Cattle Dog / Blue Heeler   
29152368  46036880  Australian Cattle Dog / Blue Heeler   
29152394  46039978                                Husky   
29152404  46041795                             Shepherd   

                breed_secondary     sex  \
20295                       NaN  Female   
23108                       NaN  Female   
25630                       NaN  Female   
27764                       NaN  Female   
28249                       NaN  Female   
...                         ...     ...   
29152364       American Bulldog    Male   
29152365       American Bulldog    Male   
29152368       American Bulldog    Male   
29152394                Terrier  Female   
29152404  Catahoula Leopard Dog  Female   

                                              description_x      id_y  \
20295     joya and faith anne are  year old spayed femal...  34418563   
23108     mollly came in with her sister annie both are ...  44990270   
25630     if you are interested in adopting please go to...  45864887   
27764     hi im teddie ann im about  months old and  lov...  46015094   
28249     daisy is available for adoption now and may be...  46033887   
...                                                     ...       ...   
29152364  hi iâ riley iâ     year old neutered male aust...  46036880   
29152365  hi iâ riley iâ     year old neutered male aust...  46041704   
29152368  hi iâ riley iâ     year old neutered male aust...  46041704   
29152394   pre adoption application must be completed fo...  46039983   
29152404  trix  is   week old beautiful female shepherdc...  46041816   

                                              description_y  \
20295     faith anne and joya are  year old spayed femal...   
23108     annie came in with her sister molly both are e...   
25630     if you are interested in adopting please go to...   
27764     hi im teddie ann im about  months old and  lov...   
28249     winnie is available for adoption now and may b...   
...                                                     ...   
29152364  hi iâ riley iâ     year old neutered male aust...   
29152365  hi iâ riley iâ     year old neutered male aust...   
29152368  hi iâ riley iâ     year old neutered male aust...   
29152394   pre adoption application must be completed fo...   
29152404  lux  is   week old beautiful female shepherdca...   

          percentuale_parole_comuni  
20295                      0.971429  
23108                      0.939394  
25630                      1.000000  
27764                      1.000000  
28249                      0.977778  
...                             ...  
29152364                   0.991342  
29152365                   1.000000  
29152368                   0.991342  
29152394                   1.000000  
29152404                   0.978495  

[5941 rows x 8 columns]

In [113]:
p = df1.loc[df1['id'] == 34418553]
p['description']

22312    joya and faith anne are  year old spayed femal...
Name: description, dtype: object

In [114]:
s = df1.loc[df1['id'] == 34418563]
s['description'] #verifico di non aver sbagliato nonostante cani con id diverso abbiano la stessa descrizione

22311    faith anne and joya are  year old spayed femal...
Name: description, dtype: object

In [115]:
query12 = result_filtrato.drop(result_filtrato[result_filtrato['sex'] == 'Unknown'].index) #rimozione dei valori che hanno sesso sconosciuto

In [116]:
query12 #dataframe finito

id_x                        breed_primary  \
20295     34418553                          Rat Terrier   
23108     44990257                          Rat Terrier   
25630     45864886                          Rat Terrier   
27764     46015088                          Rat Terrier   
28249     46033636                          Rat Terrier   
...            ...                                  ...   
29152364  46036308  Australian Cattle Dog / Blue Heeler   
29152365  46036308  Australian Cattle Dog / Blue Heeler   
29152368  46036880  Australian Cattle Dog / Blue Heeler   
29152394  46039978                                Husky   
29152404  46041795                             Shepherd   

                breed_secondary     sex  \
20295                       NaN  Female   
23108                       NaN  Female   
25630                       NaN  Female   
27764                       NaN  Female   
28249                       NaN  Female   
...                         ...     ...   
29152364       American Bulldog    Male   
29152365       American Bulldog    Male   
29152368       American Bulldog    Male   
29152394                Terrier  Female   
29152404  Catahoula Leopard Dog  Female   

                                              description_x      id_y  \
20295     joya and faith anne are  year old spayed femal...  34418563   
23108     mollly came in with her sister annie both are ...  44990270   
25630     if you are interested in adopting please go to...  45864887   
27764     hi im teddie ann im about  months old and  lov...  46015094   
28249     daisy is available for adoption now and may be...  46033887   
...                                                     ...       ...   
29152364  hi iâ riley iâ     year old neutered male aust...  46036880   
29152365  hi iâ riley iâ     year old neutered male aust...  46041704   
29152368  hi iâ riley iâ     year old neutered male aust...  46041704   
29152394   pre adoption application must be completed fo...  46039983   
29152404  trix  is   week old beautiful female shepherdc...  46041816   

                                              description_y  \
20295     faith anne and joya are  year old spayed femal...   
23108     annie came in with her sister molly both are e...   
25630     if you are interested in adopting please go to...   
27764     hi im teddie ann im about  months old and  lov...   
28249     winnie is available for adoption now and may b...   
...                                                     ...   
29152364  hi iâ riley iâ     year old neutered male aust...   
29152365  hi iâ riley iâ     year old neutered male aust...   
29152368  hi iâ riley iâ     year old neutered male aust...   
29152394   pre adoption application must be completed fo...   
29152404  lux  is   week old beautiful female shepherdca...   

          percentuale_parole_comuni  
20295                      0.971429  
23108                      0.939394  
25630                      1.000000  
27764                      1.000000  
28249                      0.977778  
...                             ...  
29152364                   0.991342  
29152365                   1.000000  
29152368                   0.991342  
29152394                   1.000000  
29152404                   0.978495  

[5940 rows x 8 columns]